connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java

```diff
@@ -356,25 +357,16 @@ public static String prunedName(PluginDesc<?> plugin) {
      * Verify whether a given plugin's alias matches another alias in a collection of plugins.
      *
      * @param alias the plugin descriptor to test for alias matching.
-     * @param plugins the collection of plugins to test against.
+     * @param aliases the collection of plugins to test against.
      * @param <U> the plugin type.
      * @return false if a match was found in the collection, otherwise true.
      */
     public static <U> boolean isAliasUnique(
             PluginDesc<U> alias,
-            Collection<PluginDesc<U>> plugins
+            Map<String, String> aliases
     ) {
-        boolean matched = false;
-        for (PluginDesc<U> plugin : plugins) {
-            if (simpleName(alias).equals(simpleName(plugin))
-                    || prunedName(alias).equals(prunedName(plugin))) {
-                if (matched) {
-                    return false;
-                }
-                matched = true;
-            }
-        }
-        return true;
+        // TODO: Mark alias collision and disable ambiguous aliases completely.
```

Comment:
We shouldn't leave this in as a TODO; we should preserve the existing behavior where ambiguous aliases are completely disabled.

Reply:
Just to clarify, the code in this PR matches the behavior on trunk, which is that some ambiguous aliases are allowed when they are ambiguous across different plugin loaders.

The TODO is for us to discuss whether to keep the trunk behavior, or change the behavior to drop these ambiguous aliases.

Reply:
Oh sorry let me clarify, it can produce ambiguous aliases if different plugin types have the same aliases.

Here's an ambiguous alias that can happen on trunk:
```
plugin-1
    package.SomethingConnector
plugin-2
    package.SomethingConverter
```

In this situation, if the iteration scans `plugin-1` first, `Something` will be aliased to `package.SomethingConnector`.
If the iteration scans `plugin-2` first, `Something` will be aliased to `package.SomethingConverter`.

I see that the proposed implementation is actually worse than trunk, which I'll need to improve. I'm still interested if you think that the ambiguity on trunk should be changed or preserved.

Reply:
Ah, gotcha--thanks for the clarification 👍

I find the ambiguity that we permit on trunk pretty distasteful, but it's also possible that there are connectors/Kafka Connect clusters out there that are working because of it.

I have no problem with eliminating that ambiguity when using service loader-based plugin discovery.

I have some qualms about eliminating it when using the existing plugin scanning-based discovery. Would it add too much implementation complexity to either preserve the existing behavior (ugly though it is), or at least emit clear error messages in exceptions that get thrown during connector startup and/or config validation?

I think I could live with it if we eliminated ambiguity altogether since the existing behavior feels like the result of an oversight and may be difficult to preserve, but since you're in the neighborhood, it'd be nice to have your thoughts on what (if anything) we can do to reasonably accommodate our users.

Reply:
I don't think that anyone should be depending on these cross-plugin ambiguous aliases, because if they were, they would be subject to spurious ClassNotFoundExceptions whenever the non-deterministic iteration order caused a plugin of the wrong type to be loaded.

I've updated this PR to make the failure deterministic, and hopefully easier to debug and remediate. Additionally, I've added a warn log message that explicitly lists the ambiguous aliases and the full class names that they refer to.

An alternative is to separate these aliases into per-plugin-type namespaces, but I don't think that is necessary at this time.

Reply:
> I don't think that anyone should be depending on these cross-plugin ambiguous aliases, because if they were, they would be subject to spurious ClassNotFoundExceptions whenever the non-deterministic iteration order caused a plugin of the wrong type to be loaded.

Plugin scanning and alias computation is currently deterministic (the various `Set<PluginDesc<?>>` objects that we use are all instances of the `TreeSet` type), so users are not subject to this risk while running a fixed version of Kafka Connect with a fixed plugin path (and fixed contents of that plugin path).

That said, I think the changes you've made are sufficient.

> An alternative is to separate these aliases into per-plugin-type namespaces, but I don't think that is necessary at this time.

Yeah, this would be nice to have but it's not a blocker. One case where it could come in handy is with plugins that implement multiple plugin interfaces (which is actually fairly common with `Converter` and `HeaderConverter` instances); you may have multiple `JsonConverter` plugins on your worker, but if only one of them implements the `Converter` interface and another implements the `Converter` and `HeaderConverter` interface, it's hard to argue that you shouldn't be able to use an alias when using the latter as a header converter.